src/test/java/com/alphawallet/attestation/demo/VerifyAttestation.sol

```diff
@@ -0,0 +1,633 @@
+        uint256(2894161621123416739138844080004799398680035544501805450971689609134516348045));
+
+    uint256 callCount;
+
+    constructor() public
+    {
+        owner = msg.sender;
+        callCount = 0;
+    }
+
+    struct Length {
+        uint decodeIndex;
+        uint length;
+    }
+
+    struct Exponent {
+        bytes riddle;
+        bytes tPoint;
+        uint256 challenge;
+    }
+    
+    //Payable variant of the attestation function for testing
+    function decodeAttestationPayable(bytes memory attestation) public returns(bool)
+    {
+        if (decodeAttestation(attestation))
+        {
+            callCount++;
+            return true;
+        }
+        else
+        {
+            revert();
+        }
+    }
+
+    //Payable variant of the attestation function for testing
+    function verifyEqualityProofPayable(bytes memory com1, bytes memory com2, bytes memory proof) public returns(bool)
+    {
+        if (verifyEqualityProof(com1, com2, proof))
+        {
+            callCount++;
+            return true;
+        }
+        else
+        {
+            revert();
+        }
+    }
+
+    function verifyEqualityProof(bytes memory com1, bytes memory com2, bytes memory proof) public view returns(bool)
+    {
+        Length memory len;
+        Exponent memory pok;
+        bytes memory attestationData;
+        uint256 decodeIndex = 0;
+
+        len = decodeLength(proof, 1);
+        decodeIndex = len.decodeIndex;
+
+        (attestationData, decodeIndex) = decodeOctetString(proof, decodeIndex);
+        pok.challenge = bytesToUint(attestationData);
+        (pok.tPoint, decodeIndex) = decodeOctetString(proof, decodeIndex);
+        
+        uint256[2] memory lhs;
+        uint256[2] memory rhs;
+        (lhs[0], lhs[1]) = extractXYFromPoint(com1);
+        (rhs[0], rhs[1]) = extractXYFromPoint(com2);
+        
+        rhs = ecInv(rhs);
+        
+        uint256[2] memory riddle = ecAdd(lhs, rhs);
+
+        bytes memory cArray = concat4Fixed(HPoint, com1, com2, pok.tPoint);
+        uint256 c = mapTo256BitInteger(cArray);
+
+        lhs = ecMul(pok.challenge, H[0], H[1]);
+        if (lhs[0] == 0 && lhs[1] == 0) { return false; } //early revert to avoid spending more gas
+
+        //ECPoint riddle multiply by proof (component hash)
+        rhs = ecMul(c, riddle[0], riddle[1]);
+        if (rhs[0] == 0 && rhs[1] == 0) { return false; } //early revert to avoid spending more gas
+
+        uint256[2] memory point;
+        (point[0], point[1]) = extractXYFromPoint(pok.tPoint);
+        rhs = ecAdd(rhs, point);
+
+        return ecEquals(lhs, rhs);
+    }
+
+    function ecEquals(uint256[2] memory ecPoint1, uint256[2] memory ecPoint2) private pure returns(bool)
+    {
+        return (ecPoint1[0] == ecPoint2[0] && ecPoint1[1] == ecPoint2[1]);
+    }
+
+    function decodeAttestation(bytes memory attestation) public view returns(bool)
```

Comment:
I am a bit unsure about this function and the flow. Is this actually expected an Attestation or a usage request? The reason I am confused is because of the proof decoding below which should not be part of an Attestation, but only an attestation usage request, but in that case the riddle is not included now.

Reply:
It should be renamed as verifyAttestationRequestProof